File: src/http/admin/listenbrainz.go
package admin

import (
	"encoding/json"
	"net/http"
	"tapesonic/listenbrainz"
	"tapesonic/model"
	"time"
)

type ListenBrainzSessionRs struct {
	Username            string
	IsScrobblingEnabled bool
	UpdatedAt           time.Time
}

func toListenBrainzSessionRs(session listenbrainz.ListenBrainzSession) ListenBrainzSessionRs {
	return ListenBrainzSessionRs{
		Username:            session.Username,
		IsScrobblingEnabled: session.IsScrobblingEnabled,
		UpdatedAt:           session.UpdatedAt.Unwrap(),
	}
}

func GetListenBrainzSession(auth *authenticator, listenbrainz *listenbrainz.ListenBrainzService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		session, err := listenbrainz.GetCurrentSession(user)
		if err != nil {
			return nil, err
		}

		if session == nil {
			return nil, model.ErrNotFound
		}

		return toListenBrainzSessionRs(*session), nil
	}
}

type CreateListenBrainzSessionRq struct {
	Token string
}

func PostListenBrainzSession(auth *authenticator, listenbrainz *listenbrainz.ListenBrainzService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		request := CreateListenBrainzSessionRq{}
		err = json.NewDecoder(r.Body).Decode(&request)
		if err != nil {
			return nil, err
		}

		session, err := listenbrainz.CreateSession(user, request.Token)
		if err != nil {
			return nil, err
		}

		return toListenBrainzSessionRs(session), nil
	}
}

func DeleteListenBrainzSession(auth *authenticator, listenbrainz *listenbrainz.ListenBrainzService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		return nil, listenbrainz.DeleteSession(user)
	}
}

type ListenBrainzSessionSettingsRq struct {
	IsScrobblingEnabled bool
}

func PutListenBrainzSessionSettings(auth *authenticator, listenbrainzSvc *listenbrainz.ListenBrainzService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		var rq ListenBrainzSessionSettingsRq
		if err := json.NewDecoder(r.Body).Decode(&rq); err != nil {
			return nil, err
		}

		session, err := listenbrainzSvc.UpdateSessionSettings(
			user,
			listenbrainz.SessionSettings{IsScrobblingEnabled: rq.IsScrobblingEnabled},
		)
		if err != nil {
			return nil, err
		}

		return toListenBrainzSessionRs(session), nil
	}
}

File: webapp/src/api/listenbrainz.ts
import api from "@/api/api";

export interface ListenBrainzSessionRs {
    Username: string;
    IsScrobblingEnabled: boolean;
    UpdatedAt: string;
}

export interface CreateListenBrainzSessionRq {
    Token: string;
}

export interface ListenBrainzSessionSettingsRq {
    IsScrobblingEnabled: boolean;
}

export default {
    async getCurrentListenBrainzSession(): Promise<ListenBrainzSessionRs | null> {
        const response = await fetch(`/api/listenbrainz/auth`, { method: "GET" });
        return await api.handleNullableResponse(response);
    },
    async createListenBrainzSession(token: string): Promise<ListenBrainzSessionRs> {
        const request: CreateListenBrainzSessionRq = {
            Token: token,
        };
        const response = await fetch(`/api/listenbrainz/auth`, { method: "POST", body: JSON.stringify(request) });
        return await api.handleRequiredResponse(response);
    },
    async deleteListenBrainzSession(): Promise<void> {
        const response = await fetch(`/api/listenbrainz/auth`, { method: "DELETE" });
        await api.handleRequiredResponse(response);
    },
    async updateListenBrainzSessionSettings(rq: ListenBrainzSessionSettingsRq): Promise<ListenBrainzSessionRs> {
        const response = await fetch(`/api/listenbrainz/settings`, { method: "PUT", body: JSON.stringify(rq) });
        return await api.handleRequiredResponse(response);
    },
}
